File: Graph/Johnson's_Algorithm.c
// Johnson's Algorithm for All-Pairs Shortest Paths
// Based on CLRS Book - Chapter 25.2

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define INF INT_MAX

struct Edge {
    int src, dest, weight;
};

struct Graph {
    int V, E;
    struct Edge* edges;
};

// ---------- Utility ----------
struct Graph* createGraph(int V, int E) {
    struct Graph* graph = (struct Graph*)malloc(sizeof(struct Graph));
    graph->V = V;
    graph->E = E;
    graph->edges = (struct Edge*)malloc(E * sizeof(struct Edge));
    return graph;
}

// ---------- Bellman–Ford ----------
int bellmanFord(struct Graph* graph, int src, int* dist) {
    int V = graph->V;
    int E = graph->E;

    for (int i = 0; i < V; i++)
        dist[i] = INF;
    dist[src] = 0;

    for (int i = 1; i <= V - 1; i++) {
        for (int j = 0; j < E; j++) {
            int u = graph->edges[j].src;
            int v = graph->edges[j].dest;
            int w = graph->edges[j].weight;
            if (dist[u] != INF && dist[u] + w < dist[v])
                dist[v] = dist[u] + w;
        }
    }

    // Check for negative cycles
    for (int j = 0; j < E; j++) {
        int u = graph->edges[j].src;
        int v = graph->edges[j].dest;
        int w = graph->edges[j].weight;
        if (dist[u] != INF && dist[u] + w < dist[v]) {
            printf("Graph contains negative weight cycle!\n");
            return 0;
        }
    }
    return 1;
}

// ---------- Dijkstra (Min-Heap Implementation) ----------
int minDistance(int dist[], int visited[], int V) {
    int min = INF, min_index = -1;
    for (int v = 0; v < V; v++)
        if (!visited[v] && dist[v] <= min) {
            min = dist[v];
            min_index = v;
        }
    return min_index;
}

void dijkstra(int** adj, int V, int src, int* result) {
    int dist[V];
    int visited[V];
    for (int i = 0; i < V; i++) {
        dist[i] = INF;
        visited[i] = 0;
    }
    dist[src] = 0;

    for (int count = 0; count < V - 1; count++) {
        int u = minDistance(dist, visited, V);
        if (u == -1) break;
        visited[u] = 1;
        for (int v = 0; v < V; v++)
            if (!visited[v] && adj[u][v] != INF && dist[u] != INF
                && dist[u] + adj[u][v] < dist[v])
                dist[v] = dist[u] + adj[u][v];
    }

    for (int i = 0; i < V; i++)
        result[i] = dist[i];
}

// ---------- Johnson’s Algorithm ----------
void johnson(struct Graph* graph) {
    int V = graph->V;

    // Step 1: Add new vertex q
    struct Graph* g2 = createGraph(V + 1, graph->E + V);
    for (int i = 0; i < graph->E; i++)
        g2->edges[i] = graph->edges[i];
    for (int i = 0; i < V; i++) {
        g2->edges[graph->E + i].src = V;
        g2->edges[graph->E + i].dest = i;
        g2->edges[graph->E + i].weight = 0;
    }

    // Step 2: Run Bellman–Ford from q
    int* h = (int*)malloc((V + 1) * sizeof(int));
    if (!bellmanFord(g2, V, h)) {
        free(h);
        return;
    }

    // Step 3: Reweight edges
    int** adj = (int**)malloc(V * sizeof(int*));
    for (int i = 0; i < V; i++) {
        adj[i] = (int*)malloc(V * sizeof(int));
        for (int j = 0; j < V; j++)
            adj[i][j] = INF;
    }

    for (int i = 0; i < graph->E; i++) {
        int u = graph->edges[i].src;
        int v = graph->edges[i].dest;
        int w = graph->edges[i].weight;
        adj[u][v] = w + h[u] - h[v];
    }

    // Step 4: Run Dijkstra from each vertex
    printf("All-Pairs Shortest Paths (Johnson's Algorithm):\n");
    for (int u = 0; u < V; u++) {
        int* dist = (int*)malloc(V * sizeof(int));
        dijkstra(adj, V, u, dist);

        printf("From vertex %d:\n", u);
        for (int v = 0; v < V; v++) {
            if (dist[v] == INF)
                printf("  to %d: INF\n", v);
            else
                printf("  to %d: %d\n", v, dist[v] - h[u] + h[v]);
        }
        printf("\n");
        free(dist);
    }

    free(h);
    for (int i = 0; i < V; i++)
        free(adj[i]);
    free(adj);
}

// ---------- Example ----------
int main() {
    int V = 5;
    int E = 8;
    struct Graph* graph = createGraph(V, E);

    graph->edges[0] = (struct Edge){0, 1, -1};
    graph->edges[1] = (struct Edge){0, 2, 4};
    graph->edges[2] = (struct Edge){1, 2, 3};
    graph->edges[3] = (struct Edge){1, 3, 2};
    graph->edges[4] = (struct Edge){1, 4, 2};
    graph->edges[5] = (struct Edge){3, 2, 5};
    graph->edges[6] = (struct Edge){3, 1, 1};
    graph->edges[7] = (struct Edge){4, 3, -3};

    johnson(graph);
    return 0;
}

File: Algorithms/Johnson's_Algorithm.f90
!===========================================================
! Johnson's Algorithm for All-Pairs Shortest Paths
! Fortran 90+
!===========================================================

module graph_module
    implicit none
    integer, parameter :: INF = 2147483647
contains

    !-------------------------------------------------------
    subroutine bellman_ford(n, m, u, v, w, h, has_negative_cycle)
        implicit none
        integer, intent(in) :: n, m
        integer, intent(in) :: u(:), v(:)
        integer, intent(in) :: w(:)
        integer, intent(out) :: h(:)
        logical, intent(out) :: has_negative_cycle
        integer :: i, j

        h = 0  ! initialize potentials

        ! Relax edges n-1 times
        do i = 1, n-1
            do j = 1, m
                if (h(u(j)) /= INF .and. h(u(j)) + w(j) < h(v(j))) then
                    h(v(j)) = h(u(j)) + w(j)
                end if
            end do
        end do

        ! Detect negative cycles
        has_negative_cycle = .false.
        do j = 1, m
            if (h(u(j)) /= INF .and. h(u(j)) + w(j) < h(v(j))) then
                has_negative_cycle = .true.
                return
            end if
        end do
    end subroutine bellman_ford

    !-------------------------------------------------------
    subroutine dijkstra(n, adj_head, adj_next, adj_v, adj_w, source, dist)
        implicit none
        integer, intent(in) :: n, source
        integer, intent(in) :: adj_head(:), adj_next(:), adj_v(:), adj_w(:)
        integer, intent(out) :: dist(:)

        logical :: visited(n)
        integer :: i, u, v, w, best, e

        dist = INF
        visited = .false.
        dist(source) = 0

        do i = 1, n
            ! Choose nearest unvisited node
            best = -1
            do u = 1, n
                if (.not. visited(u)) then
                    if (best == -1 .or. dist(u) < dist(best)) best = u
                end if
            end do

            if (dist(best) == INF) exit
            visited(best) = .true.

            ! Relax neighbors
            e = adj_head(best)
            do while (e /= 0)
                v = adj_v(e)
                w = adj_w(e)
                if (dist(best) + w < dist(v)) then
                    dist(v) = dist(best) + w
                end if
                e = adj_next(e)
            end do
        end do
    end subroutine dijkstra

end module graph_module

!===========================================================
program johnson_algorithm
    use graph_module
    implicit none

    integer :: n, m
    integer, allocatable :: u(:), v(:), w(:)
    integer, allocatable :: h(:)
    logical :: bad_cycle

    ! adjacency structure for reweighted graph
    integer, allocatable :: adj_head(:), adj_next(:), adj_v(:), adj_w(:)
    integer :: i, e

    integer, allocatable :: dist(:), all_dist(:,:)

    print *, "Enter number of vertices (n):"
    read *, n
    print *, "Enter number of edges (m):"
    read *, m

    allocate(u(m), v(m), w(m))
    allocate(h(n))
    allocate(adj_head(n), adj_next(m), adj_v(m), adj_w(m))
    allocate(dist(n))
    allocate(all_dist(n,n))

    print *, "Enter edges: u v weight"
    do i = 1, m
        read *, u(i), v(i), w(i)
    end do

    !---------------------------------------
    ! Step 1: Run Bellman–Ford from dummy node
    !---------------------------------------
    call bellman_ford(n, m, u, v, w, h, bad_cycle)

    if (bad_cycle) then
        print *, "Graph contains a NEGATIVE cycle. Johnson's algorithm cannot continue."
        stop
    end if

    !---------------------------------------
    ! Step 2: Reweight edges w' = w + h[u] - h[v]
    !---------------------------------------
    adj_head = 0
    do i = 1, m
        adj_next(i) = adj_head(u(i))
        adj_head(u(i)) = i
        adj_v(i) = v(i)
        adj_w(i) = w(i) + h(u(i)) - h(v(i))
    end do

    !---------------------------------------
    ! Step 3: Run Dijkstra for each vertex
    !---------------------------------------
    do i = 1, n
        call dijkstra(n, adj_head, adj_next, adj_v, adj_w, i, dist)

        ! Recover original distances
        do e = 1, n
            if (dist(e) < INF) then
                all_dist(i,e) = dist(e) - h(i) + h(e)
            else
                all_dist(i,e) = INF
            end if
        end do
    end do

    !---------------------------------------
    ! Print result matrix
    !---------------------------------------
    print *, ""
    print *, "All-Pairs Shortest Path Distance Matrix:"
    do i = 1, n
        do e = 1, n
            if (all_dist(i,e) == INF) then
                write(*,'(A8)', advance='no') "INF"
            else
                write(*,'(I8)', advance='no') all_dist(i,e)
            end if
        end do
        print *, ""
    end do

end program johnson_algorithm
